x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/query/RelevanceMatchQueryBuilder.java

```diff
@@ -67,31 +80,49 @@ public static RelevanceMatchQueryBuilder fromXContent(final XContentParser parse
         if (builder.query == null) {
             throw new ParsingException(parser.getTokenLocation(), "[relevance_match] requires a query, none specified");
         }
+
+        builder.setRelevanceSettingsService(relevanceSettingsService);
+
         return builder;
     }
 
     @Override
     protected void doWriteTo(final StreamOutput out) throws IOException {
         out.writeString(query);
+        out.writeOptionalString(relevanceSettingsId);
     }
 
     @Override
     protected void doXContent(final XContentBuilder builder, final Params params) throws IOException {
         builder.startObject(NAME);
 
-        builder.field(FIELD_QUERY.getPreferredName(), query);
+        builder.field(QUERY_FIELD.getPreferredName(), query);
+        if (relevanceSettingsId != null) {
+            builder.field(RELEVANCE_SETTINGS_FIELD.getPreferredName(), relevanceSettingsId);
+        }
 
         builder.endObject();
     }
 
     @Override
     protected Query doToQuery(final SearchExecutionContext context) throws IOException {
-        Collection<String> fields = queryFieldsResolver.getQueryFields(context);
-        if (fields.isEmpty()) {
-            throw new IllegalArgumentException("[relevance_match] query cannot find text fields in the index");
+
+        Collection<String> fields;
+        if (relevanceSettingsId != null) {
+            try {
+                RelevanceSettings relevanceSettings = relevanceSettingsService.getRelevanceSettings(relevanceSettingsId);
+                fields = relevanceSettings.getFields();
+            } catch (RelevanceSettingsService.RelevanceSettingsNotFoundException e) {
+                throw new IllegalArgumentException("[relevance_match] query can't find search settings: " + relevanceSettingsId);
+            }
+        } else {
+            fields = queryFieldsResolver.getQueryFields(context);
+            if (fields.isEmpty()) {
+                throw new IllegalArgumentException("[relevance_match] query cannot find text fields in the index");
+            }
```

Comment:
In my opinion, the QueryFieldResolver::getQueryFields ethod should have two parameters:
- SearchExecutionContext context
- String relevanceSettingsId

This is the responsibility of the QueryFieldsResolver to retrieve the fields from the relevanceSettings (and to load the settings).

Reply:
I’m looking at the changes to include `RelevanceSettingsService` inside `QueryFieldsResolver`. I believe my original reason to do so was not to hide `RelevanceSettingsService`, as this will be used for more than fields (weights, boosts, possibly relevance tuning).

I think it’s better to keep them separated. I can build it the other way round; `QueryFieldsResolver` could contain a `RelevanceSettingsService` instance to retrieve the fields in case a settings id is specified.

However, given that probably the query building code will end up on its own class (and will do it now since I’m working on curations), maybe it’s a little early to decide on this. I say let’s leave this as is and work on a new class that will perform the necessary query building.

WDYT?